src/pgpwde_fmt_plug.c

```diff
@@ -0,0 +1,440 @@
+	AES_cbc_encrypt(esk, out, 128, &aes_key, iv, AES_DECRYPT);
+
+	// Remove OAEP padding from ESK
+	return PKCS1oaepMGF1Unpack(out, 128, (unsigned char*)"", 0, msg, &length);
+}
+
+static void init(struct fmt_main *self)
+{
+#ifdef _OPENMP
+	static int omp_t = 1;
+	omp_t = omp_get_max_threads();
+	self->params.min_keys_per_crypt *= omp_t;
+	omp_t *= OMP_SCALE;
+	self->params.max_keys_per_crypt *= omp_t;
+#endif
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	cracked = mem_calloc(sizeof(*cracked), self->params.max_keys_per_crypt);
+	cracked_count = self->params.max_keys_per_crypt;
+}
+
+static void done(void)
+{
+	MEM_FREE(cracked);
+	MEM_FREE(saved_key);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *p = ciphertext, *ctcopy, *keeptr;
+	int extra;
+	int res;
+
+	if (strncmp(ciphertext, FORMAT_TAG, FORMAT_TAG_LENGTH))
+		return 0;
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "*")) == NULL) // version
+		goto bail;
+	if (!isdec(p))
+		goto bail;
+	if ((p = strtokm(NULL, "*")) == NULL) // symmAlg
+		goto bail;
+	if (!isdec(p))
+		goto bail;
+	res = atoi(p);
+	if (res != 9) // AES256
+		goto bail;
+	if ((p = strtokm(NULL, "*")) == NULL) // s2ktype
+		goto bail;
+	if (!isdec(p))
+		goto bail;
+	res = atoi(p);
+	if (res != 100)
+		goto bail;
+	if ((p = strtokm(NULL, "*")) == NULL) // hashIterations
+		goto bail;
+	if (!isdec(p))
+		goto bail;
+	if ((p = strtokm(NULL, "*")) == NULL) // salt
+		goto bail;
+	if (hexlenl(p, &extra) > 16 * 2 || extra)
+		goto bail;
+	if ((p = strtokm(NULL, "*")) == NULL) // ESK
+		goto bail;
+	if (hexlenl(p, &extra) > 128 * 2 || extra)
+		goto bail;
+	if (!ishexlc(p))
+		goto bail;
+
+	MEM_FREE(keeptr);
+	return 1;
+
+bail:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs;
+	int i;
+	char *p = ciphertext, *ctcopy, *keeptr;
+
+	memset(&cs, 0, sizeof(cs));
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+	ctcopy += FORMAT_TAG_LENGTH;
+	p = strtokm(ctcopy, "*");
+	cs.version = atoi(p);
+	p = strtokm(NULL, "*");
+	cs.symmAlg = atoi(p);
+	p = strtokm(NULL, "*");
+	cs.s2ktype = atoi(p);
+	p = strtokm(NULL, "*");
+	cs.hashIterations = atoi(p);
+	if (cs.hashIterations == 0)
+		cs.bytes = 1 << 16;
+	else
+		cs.bytes = 1 << cs.hashIterations;
+	p = strtokm(NULL, "*");
+	cs.salt_size = 16;
+	for (i = 0; i < cs.salt_size; i++)
+		cs.salt[i] = (atoi16[ARCH_INDEX(p[2*i])] << 4) | atoi16[ARCH_INDEX(p[2*i+1])];
+	p = strtokm(NULL, "*");
+	for (i = 0; i < 128; i++)
+		cs.esk[i] = (atoi16[ARCH_INDEX(p[2*i])] << 4) | atoi16[ARCH_INDEX(p[2*i+1])];
+
+	MEM_FREE(keeptr);
+
+	return (void *)&cs;
+}
+
+static void set_salt(void *salt)
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+
+	memset(cracked, 0, sizeof(cracked[0]) * cracked_count);
```

Comment:
Good point. This is a common code pattern in our existing code. I am now using `count` in the updated commit instead of `cracked_count`.